### PR TITLE
OFAST-66 Lesson URLs

### PR DESCRIPTION
**Pull request by AsirAAlam**
- Lessons now have their own shareable URLs
- E.g. ofast.io/learn/dp
- Added a test greedy lesson at ofast.io/learn/greedy
- Restructured the way lesson buttons are rendered in the learn page:
  - Buttons are dynamically generated from a list of lessons
  - The learn page's buttons each have a URL parameter associated with them that redirects them to the lesson page
  - Depending on the URL parameter, the lesson page displays the corresponding lesson content

https://github.com/ofast-team/ofast/assets/46213673/4b11b891-bf3f-4a5e-a42a-efb725f5a95a

### CODE REVIEW

File: src/pages/LearnPage.tsx [hunk 1]
```diff
@@ -64,101 +59,96 @@ function LessonGroup({ title, children }: LessonGroupProps) {
...
+    group: 'Intro Data Structures',
+    name: 'Lists & Vectors',
+    urlParam: 'dp',
+  },
+  {
+    group: 'Intro Data Structures',
+    name: 'Stacks & Queues',
+    urlParam: 'dp',
+  },
+]
```
Comment (cameroncuster) [inline src/pages/LearnPage.tsx:99]: we need to discuss where this metadata should live, but I think this is fine for now

File: src/pages/LearnPage.tsx [hunk 2]
```diff
@@ -64,101 +59,96 @@ function LessonGroup({ title, children }: LessonGroupProps) {
...
 
 export default function LearnPage() {
-  const [lesson, setLesson] = useState<React.ReactNode>(null)
+  const groupedLessons: {
+    [key: string]: lesson[]
+  } = {}
```
Comment (cameroncuster) [inline src/pages/LearnPage.tsx:104]: is this better style than making a `Record<string, x>` ?

File: src/pages/LearnPage.tsx [hunk 3]
```diff
@@ -64,101 +59,96 @@ function LessonGroup({ title, children }: LessonGroupProps) {
...
+        return (
+          <>
+            <LessonGroup title={lessonGroup[0].toString()}>
+              <InlineSpacing spacing={40} />
+              {lessonGroup[1].map((lesson) => (
+                <>
+                  <LessonButton>
+                    <Link
+                      style={{ color: 'inherit', textDecoration: 'none' }}
+                      to={'/learn/' + lesson.urlParam}
```
Comment (cameroncuster) [inline src/pages/LearnPage.tsx:132]: this is a super nit (I'm not picky about this), but some people prefer to use formatted strings like 
\`/learn/${lesson.urlParam}\`

File: src/pages/LessonPage.tsx [hunk 4]
```diff
@@ -74,7 +78,53 @@ function LessonBlockWrapper({
...
+const lessons = {
+  dp: [
+    <ReadingBlock content={dp} />,
+    <MCQBlock question={q1} answerOptions={ans1} correctOptions={correct1} />,
+    <MCQBlock question={q2} answerOptions={ans2} correctOptions={correct2} />,
+    <FITBBlock
+      question={
+        'Is this the real life? Is this just {fantasy}? ' +
+        'Caught in a {landslide}, no escape from reality. ' +
+        'Open your eyes, look up to the {skies} and see...'
```
Comment (cameroncuster) [inline src/pages/LessonPage.tsx:105]: nice lol

**Review by cameroncuster — APPROVED** [review]
Gave a pretty high level review here, but tried to a note a few things that jumped out at me

File: src/pages/LearnPage.tsx (hunk 1 above)
Comment (AsirAAlam) [reply src/pages/LearnPage.tsx:99]: yes, we can next meeting

**Review by gblake31 — COMMENTED** [review]
This is huge. Thanks for working on this.

File: src/pages/LearnPage.tsx (hunk 2 above)
Comment (AsirAAlam) [reply src/pages/LearnPage.tsx:104]: interesting! i was wondering what would be a good way to do this

File: src/pages/LearnPage.tsx (hunk 3 above)
Comment (AsirAAlam) [reply src/pages/LearnPage.tsx:132]: good to know! i'm going to leave it though since i don't think it makes a difference in this case

**Pull request merged**